### PR TITLE
[PRX] Add Extractors

### LINKED ISSUE #2144: PRX podcast support
### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2021.12.27**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

United States

### Example URLs

- Single audio:  https://beta.prx.org/stories/239600
- Podcast channel:  https://beta.prx.org/series/37658

### Description

PRX are a major producer of podcasts, serving over 70 million downloads per month according to https://www.prx.org/company/about/

The site is not presently supported by yt-dlp.

### Verbose log

```shell
[debug] Command-line config: ['-Uv', 'https://exchange.prx.org/pieces/239600-tpp-17-55-final-edition']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8 (No ANSI), err utf-8 (No ANSI), pref UTF-8
[debug] yt-dlp version 2021.12.01 [91f071af6]
[debug] Python version 3.10.1 (CPython 64bit) - Linux-4.4.194-perf-g4294a15-dirty-aarch64-with-libc
[debug] exe versions: ffmpeg 4.4.1 (setts), ffprobe 4.4.1
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2021.12.27, Current version: 2021.12.01
ERROR: It looks like you installed yt-dlp with a package manager, pip, setup.py or a tarball; Use that to update
[debug] [generic] Extracting URL: https://exchange.prx.org/pieces/239600-tpp-17-55-final-edition
[generic] 239600-tpp-17-55-final-edition: Requesting header
[redirect] Following redirect to https://beta.prx.org/stories/239600
[debug] [generic] Extracting URL: https://beta.prx.org/stories/239600
[generic] 239600: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] 239600: Downloading webpage
[generic] 239600: Extracting information
[debug] Looking for video embeds
ERROR: Unsupported URL: https://beta.prx.org/stories/239600
Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1329, in wrapper
    return func(self, *args, **kwargs)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 1398, in __extract_info
    ie_result = ie.extract(url)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 597, in extract
    ie_result = self._real_extract(url)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/extractor/generic.py", line 3813, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://beta.prx.org/stories/239600
```

### PR DESCRIPTION
**Pull request by coletdjnz**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [X] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Closes #2144, resolves https://github.com/ytdl-org/youtube-dl/issues/15948

Depends on #2069 for `modified_timestamp`

Adds support for stories, series, accounts, stories search and series search.

API was nice so I added more just for fun.

KNOWN ISSUES
- ~~The stories are sometimes split up into multiple mp3 files that would need to be merged together. For now we are using multi_video until yt-dlp implements a concat PP.~~ Fixed now

TODO:
- [X] testing
- [ ] could prob do with some more generalization & DRY cleanup

### CODE REVIEW

**Comment by dredmorbius** [comment]
Thanks, looking forward to trying this!

File: yt_dlp/extractor/prx.py [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+        return self._download_json(
+            urljoin('https://cms.prx.org/api/v1/', path), item_id, query=query, fatal=fatal, note=note)
+
+    @staticmethod
+    def _extract_embedded_data(response):
+        return traverse_obj(response, '_embedded', expected_type=dict, default={})
+
+    @classmethod
+    def _get_prx_embed_response(cls, response, section):
+        return cls._extract_embedded_data(response).get(f'prx:{section}')
```
Comment (pukkandan) [inline yt_dlp/extractor/prx.py:31]:
```suggestion
    def _get_prx_embed_response(response, section):
        return traverse_obj(response, ('_embedded', f'prx:{section}'))
```
I noticed _extract_embedded_data isn't used anywhere else. This probly doesnt even need to be a function though

File: yt_dlp/extractor/prx.py [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+                break
+
+            for entry_response in items:
+                res = entry_func(entry_response)
+                if res:
+                    yield res
+
+            total += response.get('count')
+            if total >= response.get('total'):
+                break
```
Comment (pukkandan) [inline yt_dlp/extractor/prx.py:151]:
```suggestion
            response = self._call_api(f'{item_id}: page {page}', endpoint, query={
                **(query or {}),
                'page': page,
                'per': 100
            })
            if not response:
                break

            items = self._get_prx_embed_response(response, 'items')
            yield from filter(None, map(entry_func, items))

            total += response['count']
            if total >= response['total']:
                break
```
for the `total += response['count']`,if the intend is to be non-fatal, you have to do `            total += response.get('count') or 0`. Just using `.get` is pointless

File: yt_dlp/extractor/prx.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+                '_type': 'video',
+                'formats': audio_pieces,
+                **info
+            }
+
+    def _real_extract(self, url):
+        story_id = self._match_id(url)
+        response = self._call_api(story_id, f'stories/{story_id}')
+        story = self._extract_story(response)
+        return story
```
Comment (pukkandan) [inline yt_dlp/extractor/prx.py:344]:
```suggestion
        return self._extract_story(response)
```
nitpick

File: yt_dlp/extractor/prx.py [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+                '_type': 'multi_video',
+                'entries': entries,
+                **info
+            }
+        else:
+            return {
+                '_type': 'video',
+                'formats': audio_pieces,
+                **info
+            }
```
Comment (pukkandan) [inline yt_dlp/extractor/prx.py:328]:
```suggestion
        audio_pieces = self._extract_audio_pieces(
            self._get_prx_embed_response(story_response, 'audio'))
        if len(audio_pieces) == 1:
            return {
                'formats': audio_pieces,
                **info
            }

        entries = [{
            **info,
            'id': '%s_part%d' % (info['id'], (idx + 1)),
            'formats': [fmt],
        } for idx, fmt in enumerate(audio_pieces)]
        return {
            '_type': 'multi_video',
            'entries': entries,
            **info
        }
```

File: yt_dlp/extractor/prx.py [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+            'url': 'https://listen.prx.org/stories/399200',
+            'only_matching': True
+        }
+    ]
+
+    def _extract_audio_pieces(self, audio_response):
+        pieces = []
+        piece_response = self._get_prx_embed_response(audio_response, 'items') or []
+        piece_response.sort(key=lambda x: int_or_none(x.get('position')))
+        for piece_response in self._get_prx_embed_response(audio_response, 'items'):
```
Comment (pukkandan) [inline yt_dlp/extractor/prx.py:289]: typo?

(also, this can be list comprehension)

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/prx.py (hunk 5 above)
Comment (coletdjnz) [reply yt_dlp/extractor/prx.py]: I have no idea what happened there :laughing:. fixed, thanks

**Pull request merged**